### DISCUSSION TITLE
Curso de OO (parte conceitual) - NotificationContext: Como funciona as notificações e como ele sabe de qual objeto é a notificação?

### DISCUSSION THREAD

**Discussion by oimarcella** [balta-io/forum#525] [Tira Dúvidas] [answered]
**Como funciona o método IsInvalid?**
Sei que ele funciona verificando se existe uma notificação para o objeto, e se existir, retorna true.

Mas elaborando a minha pergunta, quando eu adiciono uma nova notificação, como funciona essa referência entre as notificações e o objeto a qual ela pertence? Porque no construtor não existe nenhuma atribuição do tipo:
```
public List<Notifications> Notifications {get; set;}

public void AddNotification (Notification notification){
            **this**.Notifications.Add(notification);
        }
```

Eu imaginava que as notificações seriam adicionadas, e dentro do objeto course por exemplo eu veria as notificações de carreerItem. Mas fui testar e não é dessa maneira. Então parece que implicitamente, ele sabe que a atribuição é para a instância do objeto que chamou o método? É isso?

**Comment by oimarcella** [comment]
Outra dúvida, depois é falado que o pacote Flunt faz essa função. Então eu deveria remover o código feito para o NotificationContext e só após isso seguir?
Ou posso seguir com meu código do jeito que está?

**Comment by henriqueweege** [comment, marked as answer]
Oi, @mlasa! Tudo bem?
Então, acho que tem alguns gaps de conceito e por isso vou do princípio, desculpa se eu estiver falando de coisas que você já sabe. 
Acho que um bom ponto de partida, só para termos certeza que estamos na mesma página, é falar da diferença entre classe e objeto: classe é a declaração, objeto é a instância. Repara que não é por termos uma classe que não podemos ter diversos objetos do tipo dessa classe... um objeto só passa a existir quando você utiliza o "new" ;)
Dito isso, quando a gente utiliza herança estamos dizendo que a classe "A" é uma extensão da classe "B", isso quer dizer que um _objeto_ do tipo "A" tem as propriedades e métodos declarados na classe B. Portanto, quando você escreve no seu código "A.IsValid", você terá o retorno da validade desse _objeto_. Em poucas palavras: as notificações são adicionadas ao objeto, à aquela instância específica da classe. Outra instância, mesmo que herde da mesma classe base, terá outro resultado. 

Sobre o "this": [conforme essa doc](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/this) para acessar as propriedades da própria classe o C# infere o uso, a menos em casos de ambiguidade, ai é necessário.

Sobre o Flunt: eu recomendo que você altere o projeto e siga o curso. Vale muito a pena conhecer, é uma ferramenta super bacana!!

Valeeeu!

**Reply by oimarcella** [reply]
Obrigada, é sempre bom ter os conceitos reforçados.
